crates/taplo-cli/src/commands/queries.rs

```diff
@@ -93,52 +103,35 @@ impl<E: Environment> Taplo<E> {
                 }
             }
             crate::args::OutputFormat::Value => {
+                let mut buf = String::new();
+                let separator = cmd.separator.unwrap_or("\n".to_string());
                 if let Some(p) = cmd.pattern {
                     let p = p.trim_start_matches('.');
 
-                    let keys = p
+                    let nodes = p
                         .parse::<Keys>()
-                        .map_err(|err| anyhow!("invalid pattern: {err}"))?;
-
-                    let nodes = node
-                        .find_all_matches(keys, false)
+                        .and_then(|keys| node.find_all_matches(keys, false))
                         .map_err(|err| anyhow!("invalid pattern: {err}"))?;
 
                     if nodes.len() == 0 {
                         return Err(anyhow!("no values matched the pattern"));
                     }
 
-                    let mut buf = String::new();
-                    for (_, node) in nodes {
-                        buf += &extract_value(&node)?;
-                        buf += "\n";
-                    }
-                    if cmd.strip_newline {
-                        if buf.ends_with('\n') {
-                            let new_len = buf.trim_end().len();
-                            buf.truncate(new_len);
-                        }
-                    } else if !buf.ends_with('\n') {
-                        buf += "\n";
-                    }
+                    let values = nodes
+                        .map(|(_, node)| extract_value(&node, &separator))
+                        .collect::<Result<Vec<String>, _>>()?;
 
-                    stdout.write_all(buf.as_bytes()).await?;
-                    stdout.flush().await?;
+                    buf += &values.join(&separator);
                 } else {
-                    let mut buf = extract_value(&node)?;
-
-                    if cmd.strip_newline {
-                        if buf.ends_with('\n') {
-                            let new_len = buf.trim_end().len();
-                            buf.truncate(new_len);
-                        }
-                    } else if !buf.ends_with('\n') {
-                        buf += "\n";
-                    }
+                    buf = extract_value(&node, &separator)?;
```

Comment:
This seems to discard anything done before. Shouldn't this instead be the following?

```suggestion
                    buf += &extract_value(&node, &separator)?;
```

Reply:
It seems strange to append when the else statement is guaranteed to call `extract_value` once 🤔 , I'll look into making `buf` be non mutable.

Reply:
My bad, I thought you were moving the `buf` out of a loop, but it's just out of an `if`. I guess there's a few options. Here's another one:

```rust
let mut buf;
if let Some(...) = ... {
    buf = String::new();
    buf += ...
} else {
    buf = extract_value(...);
}
...
```